### ISSUE TITLE
Compatibility with MUVERA encoding

### ISSUE THREAD

**Issue by davidmezzetti** [lightonai/pylate#142]
First off, excellent work with this project and all the hard work you guys have put into late interaction models!

I've been working to add [MUVERA to txtai](https://github.com/neuml/txtai/issues/952) and I've been having difficulty getting any PyLate models to work well with MUVERA encoding.

[I also asked this question over on the muvfde repo](https://github.com/viig99/muvfde/issues/1) and used those benchmark scripts to rule out anything I'm doing wrong (results copied here for clarity).

| Metric                       | ModernColBERT        | FDE(20,5,16) ModernColBERT   |
|------------------------------|----------------------|------------------------------|
| ndcg_at_1                    | 0.35294              | 0.05882                      |
| ndcg_at_3                    | 0.31416              | 0.05294                      |
| ndcg_at_5                    | 0.28701              | 0.05297                      |
| ndcg_at_10                   | 0.26125              | 0.05245                      |
| ndcg_at_20                   | 0.24186              | 0.05263                      |
| ndcg_at_100                  | 0.247                | 0.0671                       |
| ndcg_at_1000                 | 0.33668              | 0.16921                      |
| map_at_1                     | 0.03499              | 0.00202                      |
| map_at_3                     | 0.06291              | 0.00427                      |
| map_at_5                     | 0.0735               | 0.00554                      |
| map_at_10                    | 0.08719              | 0.00736                      |
| map_at_20                    | 0.09664              | 0.00914                      |
| map_at_100                   | 0.11057              | 0.01303                      |
| map_at_1000                  | 0.12253              | 0.02048                      |
| recall_at_1                  | 0.03499              | 0.00202                      |
| recall_at_3                  | 0.07216              | 0.00647                      |
| recall_at_5                  | 0.09391              | 0.01138                      |
| recall_at_10                 | 0.1267               | 0.01995                      |
| recall_at_20                 | 0.15834              | 0.03348                      |
| recall_at_100                | 0.26555              | 0.09853                      |
| recall_at_1000               | 0.57746              | 0.4513                       |

Curious if anyone has tried MUVERA encoding and has insights into parameters that work. Or if there is something fundamentally different with these embeddings vs the ones generated with models trained with the Stanford code.

Thank you!

**Comment by NohTow** [comment]
Uh, there is 0 reasons I could think of that would make the models trained with PyLate worse for MUVERA: the objective is the exact same!
Could you quickly try [ColBERT-small](https://huggingface.co/answerdotai/answerai-colbert-small-v1)? This is the exact same training setup as GTE-ModernColBERT, the only difference is the base model.

I think I have a guess though, we agree that the results you reported for GTE-ModernColBERT without MUVERA is the one we report and not something you reproduced?
If yes, I suspect that you are running those tests on CPU but GTE-ModernColBERT inference is broken on CPU, see #125.
I really need to fix this (I have a branch working, but I want to make sure I do not miss anything for training).

**Comment by NohTow** [comment]
This is very weird.
IIUC:
- Stanford models work correctly with MUVERA, both with PyLate and stanford-nlp.
- Jina model works correctly with MUVERA, both with PyLate and stanford-nlp.
- ColBERT-small does not work with PyLate (does it with stanford?)
- GTE-ModernColBERT does not work with PyLate (we cannot test with stanford here)

What's very odd is that ColBERT-small is not working while it is trained with stanford-nlp, so it should work as well. Maybe this one is because

**Comment by NohTow** [comment]
It's very very strange, because it does not seems to be related to PyLate inference (because you manage to get good scores with some of the models) nor to be from the PyLate training (because ColBERT-small is bad as well).
Also, AFAIK all of these models are trained with KD using cross-encoders, so even there I am not sure why a particular model would be better or worse in this regard.
I know @bclavie did some tests on MUVERA (and he is the author of ColBERT-small), so maybe he has some insights as well?

**Comment by davidmezzetti** [comment]
More to the mystery here.

I trained a [PyLate model](https://huggingface.co/NeuML/pubmedbert-base-colbert) using the [standard contrastive learning example](https://github.com/lightonai/pylate?tab=readme-ov-file#contrastive-training). It performs much better with MUVERA even though this model is designed for medical data vs being a generic model.

| Metric                       | FDE(20,5,16) PubMedBERT ColBERT   |
|------------------------------|------------------------------|
| ndcg_at_1                    | 0.30805                      |
| ndcg_at_3                    | 0.28178                      |
| ndcg_at_5                    | 0.26326                      |
| ndcg_at_10                   | 0.23928                      |
| ndcg_at_20                   | 0.21769                      |
| ndcg_at_100                  | 0.21301                      |
| ndcg_at_1000                 | 0.30702                      |
| map_at_1                     | 0.03329                      |
| map_at_3                     | 0.05616                      |
| map_at_5                     | 0.06669                      |
| map_at_10                    | 0.07728                      |
| map_at_20                    | 0.0839                       |
| map_at_100                   | 0.09398                      |
| map_at_1000                  | 0.10515                      |
| recall_at_1                  | 0.03329                      |
| recall_at_3                  | 0.06608                      |
| recall_at_5                  | 0.08951                      |
| recall_at_10                 | 0.12092                      |
| recall_at_20                 | 0.14343                      |
| recall_at_100                | 0.22799                      |
| recall_at_1000               | 0.54349                      |

Is it possible that contrastive vs the knowledge distillation objective is a factor?

**Comment by NohTow** [comment]
I was wondering at some point, but AFAIK Jina-ColBERT-v2 is also further fine-tuned using KD to distill a teacher...
So it does not seems like it is the reason

**Comment by NohTow** [comment]
> Training on contrastive -> KD didn't make a difference.

That is a bummer, your intuition made a lot of sense and I actually have a good model pre-trained with contrastive before KD that could have be used to confirm this :(

IIRC you said > I trained a [PyLate model](https://huggingface.co/NeuML/pubmedbert-base-colbert) using the [standard contrastive learning example](https://github.com/lightonai/pylate?tab=readme-ov-file#contrastive-training).

When you say contrastive -> KD did not make a difference, is the model with contrastive ok and the model with KD better or worse?

> I’d like to confirm whether the [ColBERT model in PyLate format](https://huggingface.co/lightonai/colbertv2.0)
with MUVERA should achieve the same performance as the Stanford ColBERT model with MUVERA. In my experiments, the former produces significantly lower scores on my dataset.

Also interested in this. Our inference should be _very_ similar/equals to the one of stanford. I know there is a precision casting that might change the output a bit, but it shouldn't be very large, so I am curious whether it actually change the results of MUVERA a lot.

**Comment by davidmezzetti** [comment]
Do you happen to have a version of the distilled dataset using this crossencoder: https://huggingface.co/cross-encoder/ms-marco-MiniLM-L6-v2 as used in the ColBERT v2 paper?

Perhaps it's something inherent with the training dataset. I see that the version used for ColBERT v2 is a 26GB json file (https://huggingface.co/colbert-ir/colbertv2.0_msmarco_64way/tree/main)

Otherwise, I don't really have any great ideas left. When I look at the code for PyLate and stanford-nlp, the training logic looks the same for kd....more or less.

**Comment by davidmezzetti** [comment]
In the middle of training with stanford-nlp and so far it looks like it does trend better with MUVERA scoring.

One difference I notice is that the [stanford code computes in batch negatives and adds that to the loss](https://github.com/stanford-futuredata/ColBERT/blob/main/colbert/training/training.py#L125).

I'll try to see if I can add that to the loss function in PyLate just to see if it makes any difference.

**Comment by NohTow** [comment]
That is really interesting, let me sum up:

- Models trained with contrastive loss are always performing okay with MUVERA, not matter the training framework
- Using knowledge distillation with the PyLate setup makes MUVERA bad, even if starting from a (working) contrastive checkpoint

You mention that stanford-nlp does IB contrastive even when doing KD, which is correct, and that it could be the reason that it maintains correct performance even when doing KD. This is a very fair assumption given that KD alone seems to be tanking while contrastive is fine. That being said, colbert-small have been trained on the stanford-nlp codebase as well, so should be using IB negatives as well during the KD training, so except if Ben removed that during his training, it should not be this.

One common thing in all the failing setup though is that we used the same set of data for the KD (well, ColBERT-small have more data, but it's part of it). I think it'll be worth to try out:

- A model trained with another KD dataset (mining/teacher), so prolly this [one](https://huggingface.co/datasets/lightonai/ms-marco-en-bge) (as GTE is trained with the gemma one)
- A model trained with the KD dataset from stanford but with PyLate
I'll try to ask people from Jina/Ben to know about the setup

**Comment by NohTow** [comment]
Cool!

Yeah I should have the un-normalized weights, I could share on Monday! This could very well be the solution because Jina did not normalize the score, only ColBERT-small and PyLate models does!
In the meantime, did you try to use the [ms-marco-en-bge](https://huggingface.co/datasets/lightonai/ms-marco-en-bge) version? It is still normalized, but comes from a different mining/teacher!

**Comment by davidmezzetti** [comment]
To follow up, I scored a subset of [ms-marco-en-bge-gemma](https://huggingface.co/datasets/lightonai/ms-marco-en-bge-gemma) using [bge-reranker-v2-gemma](https://huggingface.co/BAAI/bge-reranker-v2-gemma). In my tests, both the MUVERA and overall scores were increasing.

I'm growing confident that normalized scores are the issue. Once the full dataset with unnormalized scores are out there, I'll confirm.

**Comment by davidmezzetti** [comment]
Added a few more of these style of models: https://huggingface.co/collections/NeuML/colbert-68cb248ce424a6d6d8277451

**Comment by NohTow** [comment]
I just saw it on Twitter!
This is so cool!!

**Comment by NohTow** [comment]
The script looks very similar to mine, so yeah I think it's more related to the model (maybe in addition to norm?) somehow

> Probably not that simple but is it possible that normalize_scores isn't set to False in the loss function?

No, it is correctly set (a mismatch between data/loss is not properly trainable AFAIK).
Thanks for your feedback, I guess we still do not have a definitive answer then... :(
